test/validation/api/crypto/odp_crypto_test_inp.c

```diff
@@ -1624,9 +1758,512 @@ static void test_auth_hashes_in_auth_range(void)
+				 ODP_CRYPTO_OP_TYPE_LEGACY,
+				 cipher,
+				 auth,
+				 AUTH_PLAINTEXT,
+				 ref,
+				 HASH_OVERLAP);
+
+	if (session == ODP_CRYPTO_SESSION_INVALID) {
+		odp_packet_free(pkt);
+		return 1;
+	}
+
+	CU_ASSERT_FATAL(hash_result_offset + ref->digest_length <= ref->length);
+
+	rc = crypto_op(pkt, &pkt, &ok, session, ODP_CRYPTO_OP_TYPE_LEGACY, 0,
+		       ref->cipher_iv, ref->auth_iv,
+		       cipher_range, auth_range,
+		       ref->aad, hash_result_offset);
+	CU_ASSERT(rc == 0);
+	CU_ASSERT(ok);
+
+	rc = odp_crypto_session_destroy(session);
+	CU_ASSERT(rc == 0);
+
+	rc = odp_packet_copy_to_mem(pkt, 0, ref->length, ref->ciphertext);
+	CU_ASSERT(rc == 0);
+
+	odp_packet_free(pkt);
+	return 0;
+}
+
+typedef struct crypto_suite_t {
+	odp_cipher_alg_t cipher;
+	odp_auth_alg_t auth;
+	alg_order_t order;
+	const odp_crypto_cipher_capability_t *cipher_capa;
+	const odp_crypto_auth_capability_t *auth_capa;
+} crypto_suite_t;
+
+/*
+ * Create test reference for combined auth & cipher by doing authentication
+ * and ciphering through separate ODP crypto operations.
+ */
+static int create_combined_ref(const crypto_suite_t *suite,
+			       crypto_test_reference_t *ref,
+			       odp_packet_data_range_t *cipher_range,
+			       odp_packet_data_range_t *auth_range,
+			       uint32_t digest_offset)
+{
+	uint32_t total_len;
+	int rc;
+	crypto_test_reference_t ref2;
+	odp_auth_alg_t first_auth, second_auth;
+	odp_cipher_alg_t first_cipher, second_cipher;
+
+	total_len = cipher_range->offset + cipher_range->length;
+	if (auth_range->offset + auth_range->length > total_len)
+		total_len = auth_range->offset + auth_range->length;
+	if (digest_offset + suite->auth_capa->digest_len > total_len)
+		total_len = digest_offset + suite->auth_capa->digest_len;
+
+	ref->cipher_key_length = suite->cipher_capa->key_len;
+	ref->cipher_iv_length = suite->cipher_capa->iv_len;
+	ref->auth_key_length = suite->auth_capa->key_len;
+	ref->auth_iv_length = suite->auth_capa->iv_len;
+	ref->digest_length = suite->auth_capa->digest_len;
+	ref->is_length_in_bits = false;
+	ref->length = total_len;
+
+	if (suite->cipher_capa->bit_mode) {
+		cipher_range->offset *= 8;
+		cipher_range->length *= 8;
+	}
+	if (suite->auth_capa->bit_mode) {
+		auth_range->offset *= 8;
+		auth_range->length *= 8;
+	}
+
+	if (ref->auth_key_length > MAX_KEY_LEN ||
+	    ref->auth_iv_length > MAX_IV_LEN ||
+	    total_len > MAX_DATA_LEN ||
+	    digest_offset + ref->digest_length > MAX_DATA_LEN)
+		CU_FAIL_FATAL("Internal error\n");
+
+	fill_with_pattern(ref->cipher_key, ref->cipher_key_length);
+	fill_with_pattern(ref->cipher_iv, ref->cipher_iv_length);
+	fill_with_pattern(ref->auth_key, ref->auth_key_length);
+	fill_with_pattern(ref->auth_iv, ref->auth_iv_length);
+	fill_with_pattern(ref->plaintext, ref->length);
+	memset(ref->plaintext + digest_offset, 0, ref->digest_length);
+
+	if (suite->order == AUTH_CIPHERTEXT) {
+		first_cipher = suite->cipher;
+		first_auth = ODP_AUTH_ALG_NULL;
```

Comment:
Above (1853, 1854) auth fields in ref gets populated from capabilities of suite->auth. 
Here the auth is set to NULL, instead of suite->auth. 
ref should get populated based on the final cipher used in operation.

Reply:
> ref should get populated based on the final auth used in operation.

I do not understand this comment and I cannot see a problem in the code. Maybe there is some confusion on what this function is doing.

This function is preparing the test reference for a combined op without using combined ops, i.e. by using cipher-only and auth-only operations to perform ciphering and authentication. So, depending on the cipher/auth order needed, the first operation is doing only ciphering or only authentication and then the second operation continues from there and processes the data further using the other operation.

Reply:
@JannePeltonen 

```
	ref->cipher_key_length = suite->cipher_capa->key_len;
	ref->cipher_iv_length = suite->cipher_capa->iv_len;
	ref->auth_key_length = suite->auth_capa->key_len;
	ref->auth_iv_length = suite->auth_capa->iv_len;
	ref->digest_length = suite->auth_capa->digest_len;
	ref->is_length_in_bits = false;
	ref->length = total_len;
```

With usage of first_auth = ODP_AUTH_ALG_NULL, do we make sure key_length, digest_size etc matches NULL's requirements? ie, digest length & key length is 0.

Reply:
> do we make sure key_length, digest_size etc matches NULL's requirements? ie, digest length & key length is 0.

That is good question and the test code does not try to ensure that the lengths etc are zero for null algorithms. Currently the code assumes that key length, digest_length, iv_length etc are ignored when the cipher/auth algorithm is ODP_CIPHER_ALG_NULL/ODP_AUTH_ALG_NULL and the API text appears to support that since it says: "ODP_CIPHER_ALG_NULL indicates that ciphering is disabled" and "When ciphering is enabled cipher key and IV need to be set". And similar for the auth case.

So, I think that either we conclude that the current API is ok but maybe needs some clarification to prevent confusion. And the test code is not buggy. Or we decide to change the API and say that key lengths etc must be set in certain way for the null algorithms and update the test accordingly.

Reply:
I tweaked the ref building code a bit and now it sets the unused auth/cipher session parameters to zeroes even if it may not be (?) strictly required by the API.

I also zero the crypto/auth range of the null algorithm now. That is not required by the API, I think, but I would actually like to make it a requirement so that non-zero range would explicitly not be allowed. In the future non-zero ranges could then be allowed with yet-to-be-defined semantics (e.g. pure data copying in the OOP op type). I think I will change the other PR to require zero ranges for null algorithms instead of saying that the ranges are to be ignored.